aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/BaseHandlerStd.java

```diff
@@ -30,4 +44,84 @@
+      final ProgressEvent<ResourceModel, CallbackContext> progress,
+      final CallbackContext callbackContext,
+      final ResourceHandlerRequest<ResourceModel> request,
+      final ProxyClient<CodeartifactClient> proxyClient,
+      final Set<String> externalConnectionsToAdd,
+      final Logger logger
+  ) {
+      ResourceModel resourceModel = request.getDesiredResourceState();
+      logger.log(String.format("Adding external connections: %s", externalConnectionsToAdd.toString()));
+      if (CollectionUtils.isNullOrEmpty(externalConnectionsToAdd)) {
+          logger.log("No external connections to add");
+          // Nothing to add, continue
+          return ProgressEvent.progress(resourceModel, callbackContext);
+      }
+
+      if (externalConnectionsToAdd.size() > 1) {
+          return ProgressEvent.failed(resourceModel, callbackContext, HandlerErrorCode.InvalidRequest,
+              "ExternalConnections for repositories are currently limited to 1.");
+      }
+
+      // Loop is currently not necessary because only 1 external connection is allowed, leaving this in for
+      // when multiple are supported
+      externalConnectionsToAdd.forEach(ec -> {
+          try {
+              AssociateExternalConnectionRequest associateExternalConnectionRequest
+                  = Translator.translateAssociateExternalConnectionsRequest(resourceModel, ec);
+
+              proxyClient.injectCredentialsAndInvokeV2(associateExternalConnectionRequest, proxyClient.client()::associateExternalConnection);
+          } catch (final AwsServiceException e) {
+              String repositoryName = progress.getResourceModel().getRepositoryName();
+              Translator.throwCfnException(e, Constants.ASSOCIATE_EXTERNAL_CONNECTION, repositoryName);
+          }
+          logger.log(String.format("Successfully associated external connection: %s", ec));
+      });
+
+      return ProgressEvent.<ResourceModel, CallbackContext>builder()
+          .resourceModel(resourceModel)
+          .status(OperationStatus.IN_PROGRESS)
+          .build();
+  }
+
+  protected ProgressEvent<ResourceModel, CallbackContext> putRepositoryPermissionsPolicy(
+      final AmazonWebServicesClientProxy proxy,
+      final ProgressEvent<ResourceModel, CallbackContext> progress,
+      final CallbackContext callbackContext,
+      final ProxyClient<CodeartifactClient> proxyClient,
+      final Logger logger
+  ) {
+      final ResourceModel desiredModel = progress.getResourceModel();
+      if (desiredModel.getPolicyDocument() == null) {
+          return ProgressEvent.progress(desiredModel, callbackContext);
+      }
+      return proxy.initiate("AWS-CodeArtifact-Repository::Update::PutRepositoryPermissionsPolicy", proxyClient, progress.getResourceModel(), progress.getCallbackContext())
+          .translateToServiceRequest(Translator::translatePutPermissionsPolicyRequest)
+          .makeServiceCall((awsRequest, client) -> {
```

Comment:
Do we always make the service call? Should we?

Maybe it doesn't matter, but based on the code I've seen if a user creates a repo using a CFN template that contains the resource policy, then they update something else about the repo in the CFN template (say they add an upstream) then when the update is applied we'll call PutRepositoryPermissionsPolicy again even though the policy is unchanged.

Does the CFN library provide a way to get a diff of the resource model (or the current/old resource model in addition to the desired one)?

Reply:
I believe there is a way to get and compare the old resource model, I'll try to update this to prevent calling this every time

Reply:
Your understanding is inline to behavior i seen while testing. I figured the optimization gain was small enough to ignore this redundancy, but I will update this to check if the permission policy had been changed before calling PutRepositoryPermissionsPolicy again.